components/camel-undertow/src/main/java/org/apache/camel/component/undertow/UndertowProducer.java

```diff
@@ -77,81 +83,135 @@ public UndertowEndpoint getEndpoint() {
+            // notify the exchange asynchronously when the HTTP exchange
+            // ends with success or failure from UndertowClientCallback
+            return false;
         }
 
-        final UndertowClientCallback clientCallback = new UndertowClientCallback(camelExchange, callback, getEndpoint(),
-            request, bodyAsByte);
+    }
 
-        // when connect succeeds or fails UndertowClientCallback will
-        // get notified on a I/O thread run by Xnio worker. The writing
-        // of request and reading of response is performed also in the
-        // callback
-        client.connect(clientCallback, uri, worker, ssl, pool, options);
+    private boolean processWebSocket(final Exchange camelExchange, final AsyncCallback camelCallback) {
+        final Message in = camelExchange.getIn();
+        try {
+            Object message = in.getBody();
+            if (!(message instanceof String || message instanceof byte[] || message instanceof Reader
+                    || message instanceof InputStream)) {
+                message = in.getBody(String.class);
+            }
 
-        // the call above will proceed on Xnio I/O thread we will
-        // notify the exchange asynchronously when the HTTP exchange
-        // ends with success or failure from UndertowClientCallback
-        return false;
+            if (message != null) {
+                final int timeout = endpoint.getSendTimeout();
+                if (isSendToAll(in)) {
+                    return webSocketHandler.send(peer -> true, message, timeout, camelExchange, camelCallback);
+                }
+                final List<String> connectionKeys = in.getHeader(UndertowConstants.CONNECTION_KEY_LIST, List.class);
+                if (connectionKeys != null) {
+                    return webSocketHandler.send(
+                        peer -> connectionKeys.contains(peer.getAttribute(UndertowConstants.CONNECTION_KEY)), message,
+                        timeout, camelExchange, camelCallback);
+                }
+                final String connectionKey = in.getHeader(UndertowConstants.CONNECTION_KEY, String.class);
+                if (connectionKey != null) {
+                    return webSocketHandler.send(
+                        peer -> connectionKey.equals(peer.getAttribute(UndertowConstants.CONNECTION_KEY)), message,
+                        timeout, camelExchange, camelCallback);
+                }
+                throw new IllegalStateException(
+                        String.format("Cannot process message which has none of the headers %s, %s or %s set: %s",
+                                UndertowConstants.SEND_TO_ALL, UndertowConstants.CONNECTION_KEY_LIST,
+                                UndertowConstants.CONNECTION_KEY, in));
+            } else {
+                /* nothing to do for a null body */
+                camelCallback.done(true);
+                return true;
+            }
+        } catch (IOException e) {
```

Comment:
Catch all kinds of exceptions, eg Exception or Throwable

Reply:
Fixed in a85b3e3